lib/Conversion/TritonGPUToLLVM/Utility.cpp

```diff
@@ -231,12 +231,30 @@ emitIndicesUsingLinearLayouts(Location loc, RewriterBase &rewriter,
       withCTAOffset ? target.getClusterCTAId(rewriter, loc) : i32_val(0);
   unsigned rank = shape.size();
   SmallVector<SmallVector<Value>> ret;
+  // Linear layout function is split in two parts below:
+  // L(r, t, w, b) = L(0, t, w, b) xor L(r, 0, 0, 0)
+  //     idxs      =    idxsBase   xor    idxsReg
+  //
+  // L(0, t, w, b) part is same for all registers, nested out of reg loop
+  //
+  // This approach produces code with lower register pressure and
+  // less computations, compared to fused L(r,t,w,b) method.
+  auto idxsBase = applyLinearLayout(loc, rewriter, *ll,
+                                    {{kRegister, i32_val(0)},
+                                     {kLane, laneId},
+                                     {kWarp, warpId},
+                                     {kBlock, blockId}});
   for (unsigned reg = 0; reg < ll->getInDimSize(str_attr("register")); reg++) {
-    auto idxs = applyLinearLayout(loc, rewriter, *ll,
-                                  {{kRegister, i32_val(reg)},
-                                   {kLane, laneId},
-                                   {kWarp, warpId},
-                                   {kBlock, blockId}});
```

Comment:
Would you be willing to try modifying applyLinearLayout to apply the xor's in the reverse order that it currently does?  (i.e. go major-to-minor rather than minor-to-major)  I think that may be an alternative way of fixing your bug.

The reason to try it this way is that otherwise we have to do this same optimization everywhere that we call applyLinearLayout.

Reply:
Yes, I am investigating `applyLinearLayout ` now. Found that it already separates "const" and "non-const: input indexes internally.
ToT computes these parts in order `[constant, non-const]`.
I've tried to switch them in [non-const, const] order, this did not improve situation.

I am surprised that LLVM can not optimize generated code, since `applyLinearLayout` separates constants from it shoud be trivial for CSE, looking into this deeped.

Reply:
> I've tried to switch them in [non-const, const] order, this did not improve situation.

Yeah that is very weird...

Reply:
Looks like CSE works as expected, at this point, I did not manage to understand why other seemingly equal changes do not give same result.

@jlebar 
Can we proceed with current solution for now?
I would like to continue investigation as a separate task.

Reply:
Go for it.